src/opf/ed.jl

```diff
@@ -0,0 +1,467 @@
+
+    model = JuMP.GenericModel{T}(optimizer)
+    model.ext[:opf_model] = EconomicDispatch  # for internal checks
+    model.ext[:solve_metadata] = Dict{Symbol,Any}(
+        :iterative_ptdf => iterative_ptdf,
+        :iterative_ptdf_tol => iterative_ptdf_tol,
+        :max_ptdf_iterations => max_ptdf_iterations,
+        :max_ptdf_per_iteration => max_ptdf_per_iteration,
+    )
+
+    # Variables
+
+    JuMP.@variable(model, pmin[g] <= pg[g in 1:G] <= pmax[g])
+    JuMP.@variable(model, rmin[g] <= r[g in 1:G] <= rmax[g])
+    JuMP.@variable(model, pf[e in 1:E])
+
+    # for upper bounds, see https://github.com/AI4OPT/OPFGenerator/pull/64/files#r1529320645
+    JuMP.@variable(model, δpb_surplus >= 0)
+    JuMP.@variable(model, δpb_shortage >= 0)
+    JuMP.@variable(model, δr_shortage >= 0)
+    JuMP.@variable(model, δf[1:E] >= 0)
+
+    JuMP.@constraint(model,
+        pf_lower_bound[e in 1:E],
+        pf[e] + δf[e] >= -pfmax[e] 
+    )
+    JuMP.@constraint(model,
+        pf_upper_bound[e in 1:E],
+        pf[e] - δf[e] <= pfmax[e] 
+    )
+
+    if !soft_power_balance
+        JuMP.set_upper_bound(δpb_surplus, 0)
+        JuMP.set_upper_bound(δpb_shortage, 0)
+    end
+
+    if !soft_reserve_requirement && minimum_reserve > 0.0
+        JuMP.set_upper_bound(δr_shortage, 0)
+    end
+
+    if !soft_thermal_limit
+        JuMP.set_upper_bound.(δf, 0)
+    end
+
+    # Constraints
+
+    JuMP.@constraint(model,
+        total_generation[g in 1:G],
+        pg[g] + r[g] <= pmax[g]
+    )
+
+    JuMP.@constraint(model,
+        power_balance,
+        sum(pg) + δpb_surplus - δpb_shortage == PD
+    )
+
+    JuMP.@constraint(model,
+        reserve_requirement,
+        sum(r) + δr_shortage >= minimum_reserve
+    )
+
+    model.ext[:PTDF] = PM.calc_basic_ptdf_matrix(data)
+    model[:ptdf_flow] = Vector{JuMP.ConstraintRef}(undef, E)
+    model.ext[:tracked_branches] = zeros(Bool, E)
+    model.ext[:ptdf_iterations] = 0
+
+    # Objective
+
+    c = [ref[:gen][g]["cost"][2] for g in 1:G]
+    q = [get!(ref[:gen][g], "reserve_cost", zeros(T, 3))[2] for g in 1:G]
+    
+    l, u = extrema(gen["cost"][1] for (i, gen) in ref[:gen])
+    (l == u == 0.0) || @warn "Data $(data["name"]) has quadratic cost terms; those terms are being ignored"
+    
+    l, u = extrema(gen["reserve_cost"][1] for (i, gen) in ref[:gen])
+    (l == u == 0.0) || @warn "Data $(data["name"]) has quadratic reserve cost terms; those terms are being ignored"
+
+    JuMP.@objective(model, Min,
+        sum(c[g]*pg[g] + ref[:gen][g]["cost"][3] for g in 1:G) +
+        sum(q[g]*r[g] + ref[:gen][g]["reserve_cost"][3] for g in 1:G) +
+        power_balance_penalty * δpb_surplus +
+        power_balance_penalty * δpb_shortage +
+        reserve_shortage_penalty * δr_shortage +
+        thermal_penalty * sum(δf)
+    )
+
+    return OPFModel{EconomicDispatch}(data, model)
+end
+
+function update!(opf::OPFModel{EconomicDispatch}, data::Dict{String,Any})
```

Comment:
I am not 100% sure about this function, namely regarding how PTDF constraints will be handled when we `update!` things.

Let's assume we have an initial instance, build it, solve it. During the solve, we add one PTDF constraint.
That PTDF constraint will have terms that depend on the nodal load values. Change the load, and the original PTDF constraint is no longer valid.

How to fix this:
1. Delete all existing PTDF constraints when we `update!` the model.
    --> this requires keeping track of JuMP-level PTDF constraints so we can delete them later. It's also the easiest to implement.
2. Re-compute the load-related coefficients of all existing PTDF constraints when we `update!`.
    --> we keep the existing JuMP constraints, but we change the RHS in-place. Possibly slow.
3. Delete PTDF constraints when we `update!`, then immediately re-add them.
   --> the difference with the above bullet is that this one does not need a separate routine to modify existing PTDF constraints. It only needs to be able to add PTDF constraints, and delete them.

Reply:
Bullets 2. and 3. above carry the risk that, if we solve a long sequence of OPF instances, calling `update!` every time, we risk ending up with a large number of PTDF constraints in the model.
Note though, this would be a performance issue, not a correctness issue.

Reply:
Good catch, I'll do 1.